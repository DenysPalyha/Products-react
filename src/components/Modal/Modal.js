import React from "react";
import PropTypes from "prop-types";
import styles from "./Modal.module.scss";

const Modal = ({ header, text, closeButton, actions, closeModale }) => {
  return (
    <div className={styles.modal} data-testid="modal-root">
      <div className={styles.bgColorModal} onClick={closeModale}></div>
      <div className={styles.container}>
        <div className={styles.headerModal}>
          <h2 className={styles.headerTitle}>{header}</h2>
          {closeButton && (
            <button className={styles.btnClose} onClick={closeModale}></button>
          )}
        </div>
        <div className={styles.modalBody}>{text}</div>
        <div className={styles.footerBtn}>{actions}</div>
      </div>
    </div>
  );
};

Modal.propTypes = {
  header: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  actions: PropTypes.node.isRequired,
  closeModale: PropTypes.func.isRequired,
  closeButton: PropTypes.bool,
};

Modal.defaultProps = {
  closeButton: true,
};

export default Modal;
