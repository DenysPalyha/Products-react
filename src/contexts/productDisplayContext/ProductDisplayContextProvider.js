import { useState } from "react";

import ProductDisplayContext from "./ProductDisplayContext";

function ProductDisplayContextProvider(props) {
  const [isProductDisplay, setIsProductDisplay] = useState(true);
  const { children } = props;

  return (
    <ProductDisplayContext.Provider
      value={{ isProductDisplay, setIsProductDisplay }}
    >
      {children}
    </ProductDisplayContext.Provider>
  );
}

export default ProductDisplayContextProvider;
