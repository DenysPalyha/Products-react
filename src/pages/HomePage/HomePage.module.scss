.heading {
  text-align: center;
  font-size: 28px;
  margin-bottom: 30px;
  color: #959595;
}

.containerBtn {
  max-width: 1140px;
  margin: 0 auto 20px;
  text-align: right;
}

.btnForModal {
  min-width: 100px;
  font-size: 15px;
  line-height: 1;
  margin: 0;
  padding: 10px;
  outline: none;
  border: 0;
  border-radius: 5px;
  background-color: #b3382c;
  cursor: pointer;
  color: #fff;

  &:first-child {
    margin-right: 10px;
  }
}
