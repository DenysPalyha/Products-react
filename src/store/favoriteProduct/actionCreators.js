import {
  SET_FAVORITE_PRODUCT,
  GET_FAVORITE_PRODUCTS,
  COUNT_FAVORITE_PRODUCT,
} from "./actions";

export const setFavoriteProductAC = (payload) => {
  return { type: SET_FAVORITE_PRODUCT, payload };
};

export const getFevoriteProducts = () => (dispatch) => {
  const favorites = JSON.parse(localStorage.getItem("favorite"));
  const countFavorite = JSON.parse(localStorage.getItem("countFavorite"));

  if (favorites !== null) {
    dispatch({ type: GET_FAVORITE_PRODUCTS, payload: favorites });
  }
  if (countFavorite !== null) {
    dispatch({ type: COUNT_FAVORITE_PRODUCT, payload: countFavorite });
  }
};

export const countFavoriteProductAC = () => {
  return { type: COUNT_FAVORITE_PRODUCT };
};
