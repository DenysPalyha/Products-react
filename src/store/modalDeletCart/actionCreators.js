import { SET_MODAL_DELETE_PRODUCT, DATA_PRODUCT_CART } from "./actions";

export const setMoadalDeleteProductAC = (payload) => ({
  type: SET_MODAL_DELETE_PRODUCT,
  payload,
});

export const dataProductCartAC = (payload) => ({
  type: DATA_PRODUCT_CART,
  payload,
});
