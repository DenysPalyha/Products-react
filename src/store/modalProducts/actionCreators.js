import { SET_MODAL_ADD_PRODUCT, DATA_PRODUCT } from "./actions";

export const setMoadalAddProducts = (payload) => ({
  type: SET_MODAL_ADD_PRODUCT,
  payload,
});

export const dataProduct = (payload) => ({
  type: DATA_PRODUCT,
  payload,
});
