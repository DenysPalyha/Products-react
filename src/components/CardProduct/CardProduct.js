import React, { useContext } from "react";
import { useSelector, useDispatch } from "react-redux";
import PropTypes from "prop-types";
import { ReactComponent as StarIcon } from "../../img/icon/star.svg";
import Button from "../Button/Button";

import {
  setMoadalAddProducts,
  dataProduct,
} from "../../store/modalProducts/actionCreators";
import {
  setFavoriteProductAC,
  countFavoriteProductAC,
} from "../../store/favoriteProduct/actionCreators";
import { setIsFavoriteProductAC } from "../../store/products/actionCreators";
import { findAndSetIsFaforiteInCart } from "../../store/productCart/actionCreators";

import styles from "./CardProduct.module.scss";
import ProductDisplayContext from "../../contexts/productDisplayContext/ProductDisplayContext";

const CardProduct = ({
  id,
  productName,
  isActiveFavorite,
  price,
  imageUrl,
  vendorCode,
  color,
}) => {
  const favorites = useSelector((store) => store.favorites.favorite);

  const { isProductDisplay } = useContext(ProductDisplayContext);

  const dispatch = useDispatch();

  return (
    <>
      {isProductDisplay ? (
        <div className={styles.container} data-testid="card-root">
          <div className={styles.wrapperHeader}>
            <p className={styles.productName}>{productName}</p>
            <Button
              textBtn={
                <StarIcon
                  className={
                    isActiveFavorite ? styles.starIconActive : styles.starIcon
                  }
                />
              }
              hendlerClik={() => {
                dispatch(
                  setFavoriteProductAC({
                    id,
                    productName,
                    price,
                    imageUrl,
                    vendorCode,
                    color,
                    isActiveFavorite,
                  })
                );
                dispatch(setIsFavoriteProductAC({ id, favorites }));
                dispatch(findAndSetIsFaforiteInCart({ id, favorites }));
                dispatch(countFavoriteProductAC());
              }}
            />
          </div>
          <div className={styles.positionImg}>
            <img
              className={styles.imagesCard}
              src={imageUrl}
              alt={productName}
            ></img>
          </div>
          <p className={styles.priceCard}>Price: {price}$</p>
          <p className={styles.codeProduct}>Vendor-code: {vendorCode}</p>
          <div className={styles.textColorProduct}>
            Color:
            <div
              className={styles.colorProduct}
              style={{ backgroundColor: color }}
            ></div>
          </div>
          <div className={styles.positionBtn}>
            <Button
              bgColor="#067305"
              textBtn="Add to cart"
              hendlerClik={() => {
                dispatch(setMoadalAddProducts(true));
                dispatch(
                  dataProduct({
                    id,
                    productName,
                    price,
                    imageUrl,
                    vendorCode,
                    color,
                    isActiveFavorite,
                  })
                );
              }}
            />
          </div>
        </div>
      ) : (
        <div className={styles.containerList} data-testid="card-root-list">
          <div className={styles.positionImg}>
            <img
              className={styles.imagesCardList}
              src={imageUrl}
              alt={productName}
            ></img>
          </div>

          <div className={styles.wrapperTextInfoList}>
            <p className={styles.productNameList}>{productName}</p>
            <p className={styles.priceCardList}>Price: {price}$</p>
            <p className={styles.codeProductList}>Vendor-code: {vendorCode}</p>
            <div className={styles.textColorProductList}>
              Color:
              <div
                className={styles.colorProductList}
                style={{ backgroundColor: color }}
              ></div>
            </div>
          </div>

          <div className={styles.wrapperHeaderList}>
            <Button
              textBtn={
                <StarIcon
                  className={
                    isActiveFavorite ? styles.starIconActive : styles.starIcon
                  }
                />
              }
              hendlerClik={() => {
                dispatch(
                  setFavoriteProductAC({
                    id,
                    productName,
                    price,
                    imageUrl,
                    vendorCode,
                    color,
                    isActiveFavorite,
                  })
                );
                dispatch(setIsFavoriteProductAC({ id, favorites }));
                dispatch(findAndSetIsFaforiteInCart({ id, favorites }));
                dispatch(countFavoriteProductAC());
              }}
            />

            <Button
              bgColor="#067305"
              textBtn="Add to cart"
              hendlerClik={() => {
                dispatch(setMoadalAddProducts(true));
                dispatch(
                  dataProduct({
                    id,
                    productName,
                    price,
                    imageUrl,
                    vendorCode,
                    color,
                    isActiveFavorite,
                  })
                );
              }}
            />
          </div>
        </div>
      )}
    </>
  );
};

CardProduct.propTypes = {
  id: PropTypes.number.isRequired,
  productName: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  imageUrl: PropTypes.string,
  vendorCode: PropTypes.number,
  color: PropTypes.string,
  isActiveFavorite: PropTypes.bool,
  favoriteCardProduct: PropTypes.func,
  addCardInCart: PropTypes.func,
};

CardProduct.defaultProps = {
  imageUrl: "./img/no_icon.png",
  vendorCode: 0,
  color: "#fff",
  favoriteCardProduct: () => {},
  addCardInCart: () => {},
  isActiveFavorite: false,
};

export default CardProduct;
