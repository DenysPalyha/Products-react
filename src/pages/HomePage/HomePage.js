import React, { useContext } from "react";
import { useSelector, useDispatch } from "react-redux";
import Button from "../../components/Button/Button";

import CardProductList from "../../components/CardProductList/CardProductList";
import Modal from "../../components/Modal/Modal";
import ProductDisplayContext from "../../contexts/productDisplayContext/ProductDisplayContext";

import { setMoadalAddProducts } from "../../store/modalProducts/actionCreators";
import {
  addProductIncartAC,
  cartCountProductAC,
} from "../../store/productCart/actionCreators";

import styles from "./HomePage.module.scss";

function HomePage() {
  const products = useSelector((store) => store.products.products);

  const isOpenModalAdd = useSelector(
    (store) => store.modalAddProducts.isOpenModalAdd
  );

  const dataProduct = useSelector(
    (store) => store.modalAddProducts.dataProduct
  );

  const { isProductDisplay, setIsProductDisplay } = useContext(
    ProductDisplayContext
  );

  const dispatch = useDispatch();

  const closeModal = () => dispatch(setMoadalAddProducts(false));

  return (
    <>
      <h2 className={styles.heading}>Products</h2>
      <div className={styles.containerBtn}>
        <Button
          bgColor="#6c7f00"
          textBtn={isProductDisplay ? "lists" : "Cards"}
          hendlerClik={() => {
            setIsProductDisplay((prevDisplay) => !prevDisplay);
          }}
        />
      </div>
      <CardProductList product={products} />

      {isOpenModalAdd && (
        <Modal
          header="Add item to cart"
          closeModale={closeModal}
          text={`Do you want to add a product: ${dataProduct.productName}?`}
          actions={
            <>
              <button
                className={styles.btnForModal}
                type="button"
                onClick={() => {
                  dispatch(addProductIncartAC(dataProduct));
                  dispatch(cartCountProductAC());
                  closeModal();
                }}
              >
                Add
              </button>
              <button
                className={styles.btnForModal}
                type="button"
                onClick={closeModal}
              >
                Cancel
              </button>
            </>
          }
        />
      )}
    </>
  );
}

export default HomePage;
