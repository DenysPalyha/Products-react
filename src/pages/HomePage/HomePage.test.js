import { render, fireEvent, screen } from "@testing-library/react";
import { Provider, useDispatch } from "react-redux";

import { store } from "../../store/index";
import ProductDisplayContextProvider from "../../contexts/productDisplayContext/ProductDisplayContextProvider";

import { setMoadalAddProducts } from "../../store/modalProducts/actionCreators";

import Modal from "../../components/Modal/Modal";
import CardProduct from "../../components/CardProduct/CardProduct";
import HomePage from "./HomePage";
import {
  addProductIncartAC,
  cartCountProductAC,
} from "../../store/productCart/actionCreators";

const ComponentOpenAndCloseModal = () => {
  const dispatch = useDispatch();

  return (
    <>
      <button
        type="button"
        onClick={() => {
          dispatch(setMoadalAddProducts(true));
        }}
      >
        Open Modal
      </button>

      <button
        type="button"
        onClick={() => {
          dispatch(
            addProductIncartAC({
              id: 1,
              productName: "Moto G Stylus 5G",
              isActiveFavorite: false,
              price: 739.99,
              imageUrl: "./img/motorola.jpg",
              vendorCode: 225739,
              color: "#366735",
            })
          );
          dispatch(cartCountProductAC());
          dispatch(setMoadalAddProducts(false));
        }}
      >
        Add
      </button>
      <button
        type="button"
        onClick={() => {
          dispatch(setMoadalAddProducts(false));
        }}
      >
        Cancel
      </button>

      <Modal
        header="modal test add card"
        text="text test add card"
        closeButton={true}
        actions={<button></button>}
        closeModale={() => {}}
      />

      <CardProduct
        id={1}
        productName="Moto G Stylus 5G"
        price={739.99}
        isActiveFavorite={true}
        color="#366735"
        imageUrl="./img/motorola.jpg"
        vendorCode={225739}
      />
    </>
  );
};

const MockedProviderHomePage = ({ children }) => (
  <ProductDisplayContextProvider>
    <Provider store={store}>
      <HomePage />
    </Provider>
  </ProductDisplayContextProvider>
);

const MockedProviderComponent = ({ children }) => (
  <ProductDisplayContextProvider>
    <Provider store={store}>
      <ComponentOpenAndCloseModal />
    </Provider>
  </ProductDisplayContextProvider>
);

describe("Home Page component", () => {
  test("should Home Page component snapshot", () => {
    const { asFragment } = render(<MockedProviderHomePage />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("checking the modal window to add a product to the cart", () => {
  test("should add a product to the cart", () => {
    render(<MockedProviderComponent />);
    const btnOpenModal = screen.getByText("Open Modal");
    fireEvent.click(btnOpenModal);

    const btnAddProduct = screen.getByText("Add");
    fireEvent.click(btnAddProduct);

    expect(screen.getByTestId("card-root")).toBeInTheDocument();
  });
});
