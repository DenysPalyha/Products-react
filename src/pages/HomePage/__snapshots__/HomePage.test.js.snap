// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page component should Home Page component snapshot 1`] = `
<DocumentFragment>
  <h2
    class="heading"
  >
    Products
  </h2>
  <div
    class="containerBtn"
  >
    <button
      class="btn"
      style="background-color: rgb(108, 127, 0);"
      type="button"
    >
      lists
    </button>
  </div>
  <div
    class="container"
  />
</DocumentFragment>
`;
